### PR TITLE
Stats suf

### PR DESCRIPTION
**Pull request by sufka861**
making an insight stats endpoint to check the ammount of request count per attribute for a given experiment id

### CODE REVIEW

File: Service/cron.job.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+            } else if (experiment.status === "active" && currentTime >= end) {
+                await ExperimentRepository.update(experiment._id, {status: "ended"})
+            }
+        }
+    })
+};
+
+          cron.schedule("0 0 1 * *", async () => {
+          await ExperimentRepository.resetMonthlyCallCount();
+         });
```
Comment (AtomicManic) [inline Service/cron.job.js:24]: I know it's not your code, but can you arrange the linting here? Didn't see it in Muhamed's PR

File: Service/route.logic.experiment.js [hunk 2]
```diff
@@ -3,23 +3,27 @@ const ffLogic = require("./feature.logic");
 const abLogic = require("./AB.test.logic");
 const Util = require("./utils");
 
+const featureFlagExpType = "f-f";
```
Comment (AtomicManic) [inline Service/route.logic.experiment.js:6]: What is this for?

File: controller/stats.controller.js [hunk 3]
```diff
@@ -3,37 +3,45 @@ const NotFoundError = require("../errors/NotFound.errors");
...
 const UserRepository = require("../repositories/user.repository");
+const GoalRepository = require("../repositories/goal.repository")
 const {mongoose} = require("mongoose");
 
 const userPercentageVariantByExperiment = async (experimentID, variant) => {
     return (await UserRepository.numUsersByExperimentIdAndVariant(experimentID, variant) / await UserRepository.numUsersByExperimentId(experimentID)) * 100;
 }
 
-const getStatistics = async (req, res) => {
+const calculateSuccessPercentage = (successCount, callCount) => {
```
Comment (AtomicManic) [inline controller/stats.controller.js:13]: where did you get success count and callCount from?

File: controller/stats.controller.js [hunk 4]
```diff
@@ -3,37 +3,45 @@ const NotFoundError = require("../errors/NotFound.errors");
...
+}
 
+const getStatistics = async (req, res) => {
     const experimentID = req.params.id;
     if (!mongoose.isValidObjectId(experimentID)) throw new ValidationError.MissingPropertyError("experiment ID");
 
     const experiment = await ExperimentRepository.retrieve(experimentID);
     if (!experiment) throw new NotFoundError.EntityNotFound(`experiment (${experimentID})`);
-    if (!experiment.call_count) throw new ServerError.ServerUnableError("calculate experiment call count");
+    if (!experiment.callCount) throw new ServerError.ServerUnableError("calculate experiment call count");
```
Comment (AtomicManic) [inline controller/stats.controller.js:23]: you can impot all the custom error individually, just for readability

File: controller/stats.controller.js [hunk 5]
```diff
@@ -62,7 +70,63 @@ const getUsersStats = async (req, res) => {
...
+    const experiment = await ExperimentRepository.retrieve(experimentId);
+    if (!experiment) throw new NotFoundError.EntityNotFound(`experiment (${experimentId})`);
+    res.status(200).send({
+        "testAttributes": experiment.testAttributes,
+        "customAttributes" : experiment.customAttributes
+    });
+}
+
+const getTestsPerMonth = async (req, res) => {
+
```
Comment (AtomicManic) [inline controller/stats.controller.js:85]: remove space

File: controller/stats.controller.js [hunk 6]
```diff
@@ -62,7 +70,63 @@ const getUsersStats = async (req, res) => {
...
+    const accountID = req.params.accountId;
+    if (!mongoose.isValidObjectId(accountID)) throw new ValidationError.MissingPropertyError("account ID");
+    const results = await ExperimentRepository.getMonthlyCalls(accountID)
+    if (!results) {
+        throw new ServerError.ServerUnableError("calculate experiment call count");
+    } else {
```
Comment (AtomicManic) [inline controller/stats.controller.js:91]: you don't really need the else here

File: controller/stats.controller.js [hunk 7]
```diff
@@ -62,7 +70,63 @@ const getUsersStats = async (req, res) => {
...
+        res.status(200).send({
+            tests: results
+        })
+    }
+}
+
+const getVariantSuccessCount = async (req, res) => {
+
```
Comment (AtomicManic) [inline controller/stats.controller.js:99]: remove space

File: controller/stats.controller.js [hunk 8]
```diff
@@ -62,7 +70,63 @@ const getUsersStats = async (req, res) => {
...
+    const goalID = req.params.goalId;
+    const experimentID = req.params.experimentId;
+    if (!mongoose.isValidObjectId(goalID)) throw new ValidationError.MissingPropertyError("goal ID");
+    if (!mongoose.isValidObjectId(experimentID)) throw new ValidationError.MissingPropertyError("experiment ID");
+    const result = await GoalRepository.getVariantSuccessCount(goalID);
+    const callCount = await ExperimentRepository.getCallCount(experimentID);
+    if(callCount === 0)
+        callCount = 1;
```
Comment (AtomicManic) [inline controller/stats.controller.js:107]: why is it one now?

File: controller/stats.controller.js [hunk 9]
```diff
@@ -62,7 +70,63 @@ const getUsersStats = async (req, res) => {
...
+    if (!result) {
+        throw new ServerError.ServerUnableError("calculate variant success count");
+    } else {
```
Comment (AtomicManic) [inline controller/stats.controller.js:110]: you dont really need the else here

File: index.js [hunk 10]
```diff
@@ -16,27 +17,33 @@ const { experimentRouter } = require("./router/experiment.router");
...
-app.use("/goal", goalRouter);
-app.use("/stats", statsRouter);
+index.use(cookieParser());
+index.use(bodyParser.urlencoded({extended:false}))
+index.use(
+  cors({
+    origin: "http://localhost:3000",
+    credentials: true,
+  })
+);
```
Comment (AtomicManic) [inline index.js:38]: can you change all the index.use to app.use?

**Review by AtomicManic — APPROVED** [review]
Good job, looks good. I asked for very minor changes, some are not your code but somehow found their way here. I'm approving, make the changes and you can merge

File: controller/stats.controller.js (hunk 3 above)
Comment (sufka861) [reply controller/stats.controller.js:13]: this is Ilays code im not sure
but callCount is in the experiment schema

File: Service/cron.job.js (hunk 1 above)
Comment (sufka861) [reply Service/cron.job.js:24]: yes

File: controller/stats.controller.js (hunk 5 above)
Comment (sufka861) [reply controller/stats.controller.js:85]: ok

File: controller/stats.controller.js (hunk 8 above)
Comment (sufka861) [reply controller/stats.controller.js:107]: not sure, its Ilays part

**Comment by sufka861** [comment]
fixed CR comments other than those on ilays code i didnt know

**Pull request merged**